File: sdk/typescript/libraries/botbuilder-solutions/src/skills/skillConversationIdFactory.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { ConversationReference, SkillConversationIdFactoryBase, Storage, SkillConversationIdFactoryOptions, SkillConversationReference, StoreItems, TurnContext } from 'botbuilder';

/**
 * A SkillConversationIdFactory that uses IStorage to store and retrieve ConversationReference instances.
 */
export class SkillConversationIdFactory extends SkillConversationIdFactoryBase {
    private readonly storage: Storage;

    public constructor(storage: Storage) {
        super();
        
        if (storage === undefined) { throw new Error('The value of storage is undefined'); }
        this.storage = storage;
    }

    public async createSkillConversationIdWithOptions(options: SkillConversationIdFactoryOptions): Promise<string> {
        if (options === undefined) { throw new Error('The value of options is undefined')};

        // Create the storage key based on the SkillConversationIdFactoryOptions
        const conversationReference: Partial<ConversationReference> = TurnContext.getConversationReference(options.activity);
        if (conversationReference === undefined) { throw new Error('The value of conversationReference is undefined'); }
        if (conversationReference.conversation === undefined) { throw new Error('The value of conversationReference.conversation is undefined'); }
        const storageKey: string = `${conversationReference.conversation.id}-${options.botFrameworkSkill.id}-${conversationReference.channelId}-skillconvo`;

        // Create the SkillConversationReference
        const skillConversationReference: SkillConversationReference = {
            conversationReference: conversationReference as ConversationReference,
            oAuthScope: options.fromBotOAuthScope
        }

        // Store the SkillConversationReference
        const skillConversationInfo: StoreItems = {} as StoreItems;
        skillConversationInfo[storageKey] = skillConversationReference;
        await this.storage.write(skillConversationInfo); 

        // Return the storageKey (that will be also used as the conversation ID to call the skill)
        return storageKey;
    }

    public async getSkillConversationReference(skillConversationId: string): Promise<SkillConversationReference>
    {
        if (skillConversationId === undefined || skillConversationId.trim().length === 0) { throw new Error('The value of skillConversationId is undefined or empty'); }

        // Get the SkillConversationReference from storage for the given skillConversationId.
        const skillConversationInfo: StoreItems = await this.storage.read([skillConversationId]);
        if (Object.keys(skillConversationInfo).length > 0) {
            const conversationInfo: SkillConversationReference = skillConversationInfo[skillConversationId];

            return conversationInfo;
        }

        throw new Error(`'skillConversationInfo' is undefined`);
    }

    public async deleteConversationReference(skillConversationId: string): Promise<void>
    {
        // Delete the SkillConversationReference from storage
        await this.storage.delete([skillConversationId]);
    }
}

File: sdk/csharp/libraries/microsoft.bot.solutions/Skills/SkillConversationIdFactory.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Bot.Schema;
using Newtonsoft.Json.Linq;

namespace Microsoft.Bot.Solutions.Skills
{
    /// <summary>
    /// A <see cref="SkillConversationIdFactory"/> that uses <see cref="IStorage"/> to store and retrieve <see cref="ConversationReference"/> instances.
    /// </summary>
    public class SkillConversationIdFactory : SkillConversationIdFactoryBase
    {
        private readonly IStorage _storage;

        public SkillConversationIdFactory(IStorage storage)
        {
            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
        }

        public override async Task<string> CreateSkillConversationIdAsync(SkillConversationIdFactoryOptions options, CancellationToken cancellationToken)
        {
            if (options == null)
            {
                throw new ArgumentNullException(nameof(options));
            }

            // Create the storage key based on the SkillConversationIdFactoryOptions
            var conversationReference = options.Activity.GetConversationReference();
            var storageKey = $"{conversationReference.Conversation.Id}-{options.BotFrameworkSkill.Id}-{conversationReference.ChannelId}-skillconvo";

            // Create the SkillConversationReference
            var skillConversationReference = new SkillConversationReference
            {
                ConversationReference = conversationReference,
                OAuthScope = options.FromBotOAuthScope,
            };

            // Store the SkillConversationReference
            var skillConversationInfo = new Dictionary<string, object> { { storageKey, JObject.FromObject(skillConversationReference) } };
            await _storage.WriteAsync(skillConversationInfo, cancellationToken).ConfigureAwait(false);

            // Return the storageKey (that will be also used as the conversation ID to call the skill)
            return storageKey;
        }

        public override async Task<SkillConversationReference> GetSkillConversationReferenceAsync(string skillConversationId, CancellationToken cancellationToken)
        {
            if (string.IsNullOrWhiteSpace(skillConversationId))
            {
                throw new ArgumentNullException(nameof(skillConversationId));
            }

            // Get the SkillConversationReference from storage for the given skillConversationId.
            var skillConversationInfo = await _storage.ReadAsync(new[] { skillConversationId }, cancellationToken).ConfigureAwait(false);
            if (skillConversationInfo.Any())
            {
                var conversationInfo = ((JObject)skillConversationInfo[skillConversationId]).ToObject<SkillConversationReference>();
                return conversationInfo;
            }

            return null;
        }

        public override async Task DeleteConversationReferenceAsync(string skillConversationId, CancellationToken cancellationToken)
        {
            // Delete the SkillConversationReference from storage
            await _storage.DeleteAsync(new[] { skillConversationId }, cancellationToken).ConfigureAwait(false);
        }
    }
}
